src/dbnode/storage/bootstrap/bootstrapper/README.md

```diff
@@ -2,12 +2,78 @@
+
+    if hasNoErrors(bootstrapResult) and allRangesFulfilled(bootstrapResult):
+        break
+```
+
 ## Bootstrappers
 
 - `fs`: The filesystem bootstrapper, used to bootstrap as much data as possible from the local filesystem.
 - `peers`: The peers bootstrapper, used to bootstrap any remaining data from peers. This is used for a full node join too.
 - `commitlog`: The commit log bootstrapper, currently only used in the case that peers bootstrapping fails. Once the current block is being snapshotted frequently to disk it might be faster and make more sense to not actively use the peers bootstrapper and just use a combination of the filesystem bootstrapper and the minimal time range required from the commit log bootstrapper.
     - *NOTE*: the commitlog bootstrapper is special cased in that it runs for the *entire* bootstrappable range per shard whereas other bootstrappers fill in the unfulfilled gaps as bootstrapping progresses.
+- `uninitialized`: The uninitialized bootstrapper, used to bootstrap a node when the whole cluster is new and there are no peers to fetch the data from.
+
+The bootstrappers satisfy the `bootstrap.Source` interface. `AvailableData()` and `AvailableIndex()` methods take shard time ranges that need to be bootstrapped and returns which ranges (possibly a subset) can be fulfilled by this bootstrapper. `Read()` method does the bootstrapping of the available data and index ranges.
+
+### Filesystem
+
+- `AvailableData()` - for each shard, reads info files (or cached data of info files) and converts block start times to available ranges
+- `AvailableIndex()` - exactly the same as `AvailableData()`
+- `Read()` - reads data ranges from info files without reading the data, then builds index segments and either flushes them to disk or keeps in memory
+
+### Commit log
+
+- `AvailableData()` - checks which shards have ever reached availability (i.e. is in `Available` or `Leaving` state) and returns the whole requested time range for those shards
+- `AvailableIndex()` - exactly the same as `AvailableData()`
+- `Read()` - for each shard reads the most recent snapshot file, and then reads commit log and checks-out all the series belonging to the namespaces that are being bootstrapped
+
+### Peer
+
+- `AvailableData()` - inspects the cluster topology and returns the whole requested time range for shards which have enough available replicas to satisfy consistency requirements
+- `AvailableIndex()` - exactly the same as `AvailableData()`
+- `Read()` - fetches shard data from peers and either persists it or checks-out into memory. Then it builds index segments and either flushes them to disk or keeps in memory
+
+### Uninitialized
+
+- `AvailableData()` - for each shard, inspects its status across peers. If the number of `Initializing` replicas is higher than `Leaving`, the shard is deemed _new_ and available to be bootstrapped by this bootstrapper
+- `AvailableIndex()` - exactly the same as `AvailableData()`
```

Comment:
This statement is the same for every bootstrapper, perhaps we should remove `AvailableIndex` (at least from the readme) as documenting it seems to carry no value?